Add IP and port allocated to the client in the response header of the CONNECT UDP listen request (Closes #12) (#18)
ietf-wg-masque/draft-ietf-masque-connect-udp-listen: draft-ietf-masque-connect-udp-listen.md

This document uses terminology from {{CONNECT-UDP}} and notational conventions
from {{!QUIC=RFC9000}}. This document uses the terms Integer and List from

> {{Section 3 of !STRUCTURED-FIELDS=RFC8941}} to specify syntax and parsing. This document uses Augmented Backus-Naur Form and parsing/serialization behvaviors from {{!ABNF=RFC5234}}

Reviewer: ```suggestion
{{Section 3 of !STRUCTURED-FIELDS=RFC8941}} to specify syntax and parsing.
This document uses Augmented Backus-Naur Form and parsing/serialization
behaviors from {{!ABNF=RFC5234}}
```

for the connect-udp-listen header field value, but future documents might define
parameters. Receivers MUST ignore unknown parameters.

# The proxy-public-address Response Header Field {#response}
Upon accepting the request, the "proxy-public-address" response header,

> MUST contain the IP and the UDP port which are allocated to the client by the

Reviewer: Do we want to make this mandatory? I was imagining this being optional

Reviewer 2: In what situation do you imagine a proxy would not want to send this? I'd assume that nobody really cares about these ~20 bytes more.

> proxy in the format {IP-Address}:{UDP-Port}

Reviewer: We should support sending multiple addresses, as it's common for a proxy to have multiple addresses

> proxy in the format {IP-Address}:{UDP-Port}

Reviewer: We'll want to define the syntax more formally. You can use [9484](https://www.rfc-editor.org/rfc/rfc9484#name-uri-template-variable-forma) as an example

Author: Done, lmk what you think of the new definition

> 

Reviewer: Another thought is about address changes mid-stream. We could support those by using a capsule instead of a header, but that might be overkill. Maybe we should acknowledge the limitation though

Reviewer 3: An address change would break connectivity, right? If I've told other devices my "public" IP through the proxy and that changes, I need to start over. That should be a different request.

Reviewer: Not necessarily. If a new address is added, then the layer above UDP could conceivably communicate that, and then migrate traffic over to using it.

But this is probably too niche and not worth the complexity.

Reviewer: Can we also add a sentence here to mention the fact that addresses cannot be updated mid-stream? Perhaps "Note that since the addresses are conveyed in HTTP response headers, a subsequent change of addresses on the proxy cannot be conveyed to the client."

for the connect-udp-listen header field value, but future documents might define
parameters. Receivers MUST ignore unknown parameters.

# The proxy-public-address Response Header Field {#response}

> Upon accepting the request, the proxy MUST allocate either one or both of IPv4address to port and IPv6address to port pairings. IPv4address, IPv6address and port are defined in {{!URI=RFC3986}}

Reviewer: nit: the prose should discuss concepts (like "IPv4 address") as opposed to the ABNF encoding (like "IPv4address").

```suggestion
Upon accepting the request, the proxy MUST select at least one public IP address to bind. For each selected address, it MUST select an open port to bind to this request. From then and until the tunnel is closed, the proxy SHALL send packets received on these IP-port tuples to the client. The proxy then MUST communicate the selected addresses and ports to the client using the "Proxy-Public-Address" header. The format of that header is defined below using IPv4address, IPv6address and port from {{Section 3.2 of !URI=RFC3986}}.
```

> 

Reviewer: Follow max line length. You should get a warning when building for these lines

> The proxy must then respond with the allocated IP and port pairings using the "proxy-public-address" response header defined as follows:

Reviewer: avoid lowercase must

> The proxy must then respond with the allocated IP and port pairings using the "proxy-public-address" response header defined as follows:

Reviewer: use camelcase when discussing headers in prose, similar to Connect-UDP-Listen

~~~ ascii-art

> proxy-public-address = IP-port-pairing (, IP-port-pairing)

Reviewer: This encoding allows sending two addresses of the same address family, which doesn't really make sense. That might be a pain to encode in the ABNF format but we can fix that with a MUST

> IP-port-pairing = IPv4address:port / IPv6address:port

Reviewer: This encoding is tricky because IPv6 addresses have colons in them. I would suggest doing something closer to RFC3986 host. Also, the ABNF format from that RFC requires double quotes to encode strings like colons.

```suggestion
IP-port-pairing = ( IPv4address / "[" IPv6address "]" ) ":" port
```

for the connect-udp-listen header field value, but future documents might define
parameters. Receivers MUST ignore unknown parameters.

# The Proxy-Public-Address Response Header Field {#response}

Upon accepting the request, the proxy MUST select at least one public IP
address to bind. For each selected address, it MUST select an open port to
bind to this request. From then and until the tunnel is closed, the proxy
SHALL send packets received on these IP-port tuples to the client. The proxy
then MUST communicate the selected addresses and ports to the client using
the "Proxy-Public-Address" header. The format of that header is defined below
using IPv4address, IPv6address and port from {{Section 3.2 of !URI=RFC3986}}.

The proxy MUST then respond with the allocated IP and port pairings using the
Proxy-Public-Address response header defined as follows:

~~~ ascii-art
proxy-public-address = IP-port-pairing (, IP-port-pairing)
IP-port-pairing = ( IPv4address / "[" IPv6address "]" ) ":" port
~~~
{: #target-format title="Proxy Address Format"}

> The proxy MUST use one each of IPv4address and IPv6address when
> proxy-public-address is defined as two IP-port-pairings.

Reviewer: ```suggestion
If the proxy sends two IP-port pairings, they MUST be of different
address families. If the client receives a Proxy-Public-Address field
with two IP-port pairings of the same address family, it MUST treat
the response as malformed. Similarly, if the field is missing or
contains a number of pairings different from one or two, it MUST
treat the response as malformed.
```

for the Connect-UDP-Listen header field value, but future documents might define
parameters. Receivers MUST ignore unknown parameters.

# The Proxy-Public-Address Response Header Field {#response}

Upon accepting the request, the proxy MUST select at least one public IP

> address to bind. For each selected address, it MUST select an open port to

Reviewer: ```suggestion
address to bind. The proxy MAY also select a second address from a different address
family. For each selected address, it MUST select an open port to
```

for the Connect-UDP-Bind header field value, but future documents might define
parameters. Receivers MUST ignore unknown parameters.

# The Proxy-Public-Address Response Header Field {#response}

Upon accepting the request, the proxy MUST select at least one public IP
address to bind. The proxy MAY also select a second address from a different address
family. For each selected address, it MUST select an open port to
bind to this request. From then and until the tunnel is closed, the proxy
SHALL send packets received on these IP-port tuples to the client. The proxy
then MUST communicate the selected addresses and ports to the client using

> the "Proxy-Public-Address" header. The format of that header is defined below

Reviewer 3: Rather than using a new header, can we add a parameter to the existing proxy-status header?

Reviewer 4: Do we have a precedence of a Proxy-Status parameter being a MUST send? I could be incorrect, but I had the impression that Proxy-Status provided visibility rather than required behavior.

Not that I have a strong opinion, but to me use of a different field sounds more natural considering that the address MUST be included in the response.

Reviewer 5: I'd be curious to hear what the benefit would be? If I understood your comment in the session correctly it sounds like you were suggesting it would be less confusing? But I am having a hard time differentiating it from "same outcome, different spelling"?

Reviewer 3: At the session, there was confusion about this new proposed header being about "here are all of the proxy's public addresses", which it is not (it is "here are the addresses I'm using for this one request").

Proxy-Status is an established header that already tells the client about what the proxy did related to this one request, and already gives you remote IP addresses and other information. So adding this to there makes sense.

Regarding if it needs to be sent... I think it's OK for this document to say "if udp-bind clients need to get the local address, then this header needs to be sent"

Author: How do you determine if the bind clients need the address? That would have to be a flag in the initial client request

Reviewer 3: No, I mean that the spec just says "clients need this, so you have to send it to them"

using IPv4address, IPv6address and port from {{Section 3.2 of !URI=RFC3986}}.

The proxy MUST then respond with the allocated IP and port pairings using the
Proxy-Public-Address response header defined as follows:

~~~ ascii-art
proxy-public-address = IP-port-pairing (, IP-port-pairing)
IP-port-pairing = ( IPv4address / "[" IPv6address "]" ) ":" port
~~~
{: #target-format title="Proxy Address Format"}

If the proxy sends two IP-port pairings, they MUST be of different

> address families. If the client receives a Proxy-Public-Address field

Reviewer 6: I think it's probably a mistake to limit this to one IP of each family.  There's nothing wrong with a proxy having multiple IPs of the same family on different network interfaces, in which case it would be optimal to allocate ports on all of them and allow ICE to find the best one.

Reviewer 7: No strong opinion on doing this, but relaxing the limits feels like it would make parsing the field a whole lot easier.

Reviewer: Thinking through this some more, maybe it's best to not restrict to two? I don't remember what the motivation was for the restriction

for the Connect-UDP-Bind header field value, but future documents might define
parameters. Receivers MUST ignore unknown parameters.

# The Proxy-Public-Address Response Header Field {#response}

Upon accepting the request, the proxy MUST select at least one public IP
address to bind. The proxy MAY also select a second address from a different address
family. For each selected address, it MUST select an open port to
bind to this request. From then and until the tunnel is closed, the proxy
SHALL send packets received on these IP-port tuples to the client. The proxy
then MUST communicate the selected addresses and ports to the client using
the "Proxy-Public-Address" header. The format of that header is defined below
using IPv4address, IPv6address and port from {{Section 3.2 of !URI=RFC3986}}.

The proxy MUST then respond with the allocated IP and port pairings using the
Proxy-Public-Address response header which is defined as a List of Strings.
The list contains a maximum of one IPv4 address and port, and one IPv6 address
and port each. IP-literal, IPv4address and Port are defined in
{{Section 3 of !URI=RFC3986}}

~~~ ascii-art
proxy-public-address = List of IP:Port pairings

> IP:Port pairing = IP-literal:Port / IPv4address:Port

Reviewer: This comment was marked as resolved but wasn't addressed (probably GitHub bug)

~~~

> {: #target-format title="Proxy Address Format"}

Reviewer: Speaking of which we should reference ABNF=RFC5234

Author: > Thinking through this some more, maybe it's best to not restrict to two? I don't remember what the motivation was for the restriction

The reason was up to one for IPv4 and one for IPv6

Author: I thought so too but github isn't letting me reply to that comment for some weird reason

Reviewer: What I meant was: I don't remember why we restricted to one IPv4 and one IPv6

Author: I mean I guess you could have more than one of each (to work with different subnets?). But then we need a way for the client to know which IP:Port would be used for what target

Reviewer 6: The client doesn't need to know.  Even the proxy may not know!  On many operating systems, the choice of outgoing interface is controlled by the routing table and is not straightforward for application software to control.

Author: Would having multiple addresses for the same family be counterproductive to this extensions since the client can no longer guarantee the IP:Port at which it will be sending/receiving.

Reviewer 6: For peer-to-peer use cases like WebRTC, it's normal to have multiple addresses of the same family.  Each address becomes an "ICE candidate", and the ICE procedure tries all pairs of addresses (in some prioritized order) to find a working connection (accounting for interference by NATs, etc.).

For client-server use cases like HTTP/3, the situation is perhaps more interesting.  If the server is trying to use this protocol, it likely does need a single consistent address for each connection.  It also needs other guarantees beyond the proxy itself, such as requiring that there are no firewalls or NATs preventing inbound connections to the ports it has opened.

My recommendation would be to remove any restriction on the number of addresses within the protocol, but note that proxies intended for client-server use should select at most one address of each family.

            <--------  STREAM(44): HEADERS
                         :status = 200
                         capsule-protocol = ?1

>                          proxy-public-address = 192.10.11.45:3030

Reviewer: IPv4 examples need to use the prefixes reserved for documentation by RFC 5737. I'd also add IPv6. The documentation prefix for v6 is 2001:db8::/32. I'd also use an ephemeral port. All together:

```suggestion
                         proxy-public-address = 192.0.2.45:54321,[2001:db8::1234]:54321
```

            <--------  STREAM(44): HEADERS
                         :status = 200
                         capsule-protocol = ?1

>                          proxy-public-address = 192.0.2.45:54321,

Reviewer: ```suggestion
                         proxy-public-address = 192.0.2.45:54321,  \
```
